pysmt/rewritings.py

```diff
@@ -832,3 +832,29 @@ def disjunctive_partition(formula):
                 to_process += cur.args()
             else:
                 yield cur
+
+
+def propagate_toplevel(formula, env=None, do_simplify=True):
+    """ Propagates the toplevel definitions and returns an equivalent formula.
+    It considers two kinds of definitions:
+    1) variable = constant
+    2) variable = variable
+    """
+    if env is None:
+        import pysmt.environment
+        env = pysmt.environment.get_env()
+    mgr = env.formula_manager
+    sigma = {}
+    for c in conjunctive_partition(formula):
+        if c.is_equals():
+            l, r = sorted(c.args(), key=hash)
+            if l.is_symbol() and (r.is_symbol() or r.is_constant()) and\
+               l not in sigma:
```

Comment:
What happens when we have a formula like `phi & a = b & a = c` assuming that hash(a) < hash(b) < hash(c)?

It seems to me that sigma will only contain a -> b as a substitution and not c -> b. I think you need a data structure to keep the equivalence classes instead of a simple `dict` and then you need to select a representative and substitute all the members of the same class with the representative....

Reply:
You are right, it will not catch the case you mentioned.

Actually, I wanted to start from a very basic one and then later improve it.

What do you think? 
Should we go for better version with some equivalence classes implementation?

Reply:
I think it should be rather simple to use a UnionFind to construct equivalence classes and then rewrite the formula. You can have a look [https://stackoverflow.com/a/3067672](here) for a simple dictionary-based UnionFind implementation.

Instead, [http://code.activestate.com/recipes/215912-union-find-data-structure/](here) there is a proper, efficient implementation. A good idea could be to put a generic UnionFind implementation in a `util` package in pysmt for possible other uses...

Reply:
I would go with the simple implementation, and put it directly in the module where this is used. AFAIK there has no other place where we use this, and if the needs arises, we can extract that.

If performance is a problem (I would assume this is the case for expressions with a huge number of equalities), we can try to cythonize the implementation similarly to what we do for the SMT-LIB parser, or identify an external library that implements this in C.

Reply:
OK. I am going to go for the basic implementation, as @marcogario suggested.